### PR TITLE
Aggregation

### LINKED ISSUE #108: support OpenTSDB style aggregation
OpenTSDB allows [aggregation](http://opentsdb.net/docs/build/html/user_guide/query/aggregators.html) of multiple series into a single one, as well as aggregation of values resulting from downsampled values. While Timely supports downsampling, it would be nice if it were also possible to combine multiple series into a single one via aggregation.

### PR DESCRIPTION
**Pull request by ivakegg**
This is to resolve #108 by implementing the Aggregation across series concept.
Changes:
  1) The list of aggregators now contains one called "none".  Selecting this aggregator will in effect disable aggregation.  This also goes for the downsampling.
  2) There is now an AggregationIterator which will aggregate across series.  It assumes that the DownsampleIterator is executed below/before it.
  3) The Downsample class was essentially refactored to be the Aggregation class and the new Downsample class extends Aggregation.
  4) The DataSourceImpl now will add the AggregationIterator after the Downsample iterator iff aggregation is enabled (something other than "none" is selected).

### CODE REVIEW

File: server/src/main/java/timely/sample/iterators/AggregationIterator.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Override
+    public boolean hasTop() {
+        // for each set of downsamples (one per subquery)
+        while (super.hasTop()) {
+            last = super.getTopKey();
+
+            // decode the set of samples (or series)
+            Map<Set<Tag>, Downsample> samples = null;
+            try {
```
Comment (dlmarion) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: Is there any reason not to include lines 78 - 88 in this try/catch block?

Comment (dlmarion) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: same question for init() above

File: server/src/main/java/timely/sample/iterators/AggregationIterator.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    public void next() throws IOException {
+        last = null;
+    }
+
+    public static void setAggregationOptions(IteratorSetting is, Map<String, String> tags, String classname) {
+        is.addOption(TAGS, serializeTags(tags));
+        is.addOption(AGGCLASS, classname);
+    }
+
+    public static String serializeTags(Map<String, String> tags) {
```
Comment (dlmarion) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: I'm curious if we should add this method to TagLIstParser.

File: server/src/main/java/timely/sample/iterators/AggregationIterator.java (hunk 1 above)
Comment (ivakegg) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: ecause only the lines included in the try actually throw the exceptions being caught....am I missing something here?

File: server/src/main/java/timely/sample/iterators/AggregationIterator.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    private static String unescape(String tag) {
+        try {
+            return URLDecoder.decode(tag, "UTF8");
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static Set<Tag> deserializeTags(String tags) {
```
Comment (dlmarion) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: Should this be an update to TagListParser.parse()?

File: server/src/main/java/timely/sample/iterators/AggregationIterator.java (hunk 2 above)
Comment (ivakegg) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: That sounds perfectly reasonable....shall do

File: server/src/main/java/timely/sample/iterators/AggregationIterator.java (hunk 3 above)
Comment (ivakegg) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: I will work this tag list parsing into the TagListParser....good stuff

File: server/src/main/java/timely/sample/iterators/AggregationIterator.java (hunk 1 above)
Comment (ivakegg) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: What I am missing is the possible performance increases gained by avoiding multiple try-catch blocks.  Perhaps minimal but measurable in any case.

Comment (ivakegg) [inline server/src/main/java/timely/sample/iterators/AggregationIterator.java]: I can fix the init, but not here as the IOException may be thrown for two separate reasons....

**Pull request merged**